### PR TITLE
Empty catch block

### PR DESCRIPTION
**Pull request by eeree**
As https://github.com/phpmd/phpmd/pull/408 stuck for some time, I've decided to finish the work started by https://github.com/greg0ire and finally push this _LIVE_.

### CODE REVIEW

**Comment by eeree** [comment]
@ravage84 
Ready for review.

File: src/main/resources/rulesets/design.xml [hunk 1]
```diff
@@ -225,4 +225,40 @@ class SuspectCode {
...
+            <![CDATA[
+Usually empty try-catch is a bad idea because you are silently swallowing an error condition
+and then continuing execution. Occasionally this may be the right thing to do, but often
+it's a sign that a developer saw an exception, didn't know what to do about it,
+and so used an empty catch to silence the problem.
+            ]]>
+        </description>
+        <priority>2</priority>
+        <properties>
+            <property name="unwanted-functions" value="var_dump,print_r,debug_zval_dump,debug_print_backtrace" description="Comma separated list of suspect function images." />
```
Comment (ravage84) [inline src/main/resources/rulesets/design.xml]: "Comma separated list of suspicious functions"

File: src/main/resources/rulesets/design.xml [hunk 2]
```diff
@@ -225,4 +225,40 @@ class SuspectCode {
...
+        </properties>
+        <example>
+            <![CDATA[
+class SuspectCode {
+
+    public function doSomething(array $items)
```
Comment (ravage84) [inline src/main/resources/rulesets/design.xml]: I don't see how this example is related? 😕 

Why don't you use [this](https://github.com/phpmd/phpmd/pull/476/files#diff-1411bd649c5b963e2926438bf50b5f3dR197)?

File: src/site/rst/rules/design.rst [hunk 3]
```diff
@@ -184,6 +184,34 @@ This rule has the following properties:
  unwanted-functions                  var_dump,print_r,debug_zval_dump,debug_print_backtrace  Comma separated list of suspect function images. 
 =================================== =============== ==================================================
 
+EmptyCatchBlock
+=======================
```
Comment (ravage84) [inline src/site/rst/rules/design.rst]: Same length , please

File: src/site/rst/rules/design.rst [hunk 4]
```diff
@@ -184,6 +184,34 @@ This rule has the following properties:
...
+
+Example: ::
+
+  class Foo {
+
+      public function bar()
+      {
+          try {
+              // ...
+          } catch (Exception $e) {}
```
Comment (ravage84) [inline src/site/rst/rules/design.rst]: Add a comment AFTER the closing bracket ``// Empty catch block``

File: src/site/rst/rules/design.rst [hunk 5]
```diff
@@ -184,6 +184,34 @@ This rule has the following properties:
...
+      }
+  }
+
+This rule has the following properties:
+
+=================================== =============== ==================================================
+ Name                                Default Value   Description
+=================================== =============== ==================================================
+ unwanted-functions                  var_dump,print_r,debug_zval_dump,debug_print_backtrace  Comma separated list of suspect function images.
```
Comment (ravage84) [inline src/site/rst/rules/design.rst]: Adjust description: "Comma separated list of suspicious functions"

File: src/test/php/PHPMD/Rule/Design/EmptyCatchBlockTest.php [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+ * @copyright 2008-2014 Manuel Pichler. All rights reserved.
+ * @license   http://www.opensource.org/licenses/bsd-license.php BSD License
+ */
+
+namespace PHPMD\Rule\Design;
+
+use PHPMD\AbstractTest;
+
+/**
+ * Test case for the {@link \PHPMD\Rule\Design\EmptyCatchBlock} class.
```
Comment (ravage84) [inline src/test/php/PHPMD/Rule/Design/EmptyCatchBlockTest.php]: Do not link like this.

You could use ``@see \PHPMD\Rule\Design\EmptyCatchBlock`` at the end of this dock block instead.

File: src/test/php/PHPMD/Rule/Design/EmptyCatchBlockTest.php [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+ *
+ * @author    Grégoire Paris <postmaster@greg0ire.fr>
+ * @copyright 2008-2014 Manuel Pichler. All rights reserved.
+ * @license   http://www.opensource.org/licenses/bsd-license.php BSD License
+ *
+ * @covers \PHPMD\Rule\Design\EmptyCatchBlock
```
Comment (ravage84) [inline src/test/php/PHPMD/Rule/Design/EmptyCatchBlockTest.php]: I think it should be ``@coversDefaultClass``, instead.

File: src/test/php/PHPMD/Rule/Design/EmptyCatchBlockTest.php [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+ * @group phpmd
```
Comment (ravage84) [inline src/test/php/PHPMD/Rule/Design/EmptyCatchBlockTest.php]: Remove these, please

File: src/test/resources/files/Rule/Design/EmptyCatchBlock/testRuleAppliesToNonStandardExceptions.php [hunk 9]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+function testRuleAppliesToNonStandardExceptions()
```
Comment (ravage84) [inline src/test/resources/files/Rule/Design/EmptyCatchBlock/testRuleAppliesToNonStandardExceptions.php]: Isn't this also about nested try/catch blocks?

File: src/test/resources/files/Rule/Design/EmptyCatchBlock/testRuleNotAppliesToCatchBlockWithComments.php [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    }
+    try {
+        // ...
+    } catch (Exception $e) {
+        /** valid line of code */
+    }
+    try {
+        // ...
+    } catch (Exception $e) {
+        /** /** valid line of code */
```
Comment (ravage84) [inline src/test/resources/files/Rule/Design/EmptyCatchBlock/testRuleNotAppliesToCatchBlockWithComments.php]: Nice!

File: src/test/resources/files/Rule/Design/EmptyCatchBlock/testRuleNotAppliesToCatchBlockWithComments.php [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+    }
+    try {
+        // ...
+    } catch (Exception $e) {
+        # valid line of code
```
Comment (ravage84) [inline src/test/resources/files/Rule/Design/EmptyCatchBlock/testRuleNotAppliesToCatchBlockWithComments.php]: I would have not thought of that, nice!

File: src/test/resources/files/Rule/Design/EmptyCatchBlock/testRuleWorksWithNestedTryCatchBlocks.php [hunk 12]
```diff
@@ -0,0 +1,18 @@
+<?php
+
+function testRuleWorksWithNestedTryCatchBlocks()
```
Comment (ravage84) [inline src/test/resources/files/Rule/Design/EmptyCatchBlock/testRuleWorksWithNestedTryCatchBlocks.php]: You are not using this, it seems. 😕

File: src/test/resources/files/Rule/Design/EmptyCatchBlock/testRuleAppliesToNonStandardExceptions.php (hunk 9 above)
Comment (ravage84) [inline src/test/resources/files/Rule/Design/EmptyCatchBlock/testRuleAppliesToNonStandardExceptions.php]: Also see ``testRuleWorksWithNestedTryCatchBlocks()``

File: src/site/rst/rules/design.rst (hunk 5 above)
Comment (ravage84) [inline src/site/rst/rules/design.rst]: This list of untwanted functions is not checked in the code, or am I wrong?

File: src/main/resources/rulesets/design.xml (hunk 2 above)
Comment (eeree) [inline src/main/resources/rulesets/design.xml]: Copy/Paste from another description. You've got me :(

File: src/main/php/PHPMD/Rule/Design/EmptyCatchBlock.php [hunk 13]
```diff
@@ -0,0 +1,74 @@
+<?php
+/**
+ * This file is part of PHP Mess Detector.
+ *
+ * Copyright (c) 2008-2012, Manuel Pichler <mapi@phpmd.org>.
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Design/EmptyCatchBlock.php]: Could you please update the file header doc block with the new one from  #479?

**Comment by eeree** [comment]
@ravage84 
All done!

File: src/main/php/PHPMD/Rule/Design/EmptyCatchBlock.php [hunk 14]
```diff
@@ -0,0 +1,74 @@
+<?php
+/**
+ * This file is part of PHP Mess Detector.
+ *
+ * Copyright (c) 2008-2017, Manuel Pichler <mapi@phpmd.org>.
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Design/EmptyCatchBlock.php]: Doc block still the old one.

File: src/main/php/PHPMD/Rule/Design/EmptyCatchBlock.php [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+
+use PHPMD\AbstractNode;
+use PHPMD\AbstractRule;
+use PHPMD\Rule\FunctionAware;
+use PHPMD\Rule\MethodAware;
+
+/**
+ * This rule class detects empty catch blocks
+ *
+ * @author    Grégoire Paris <postmaster@greg0ire.fr>
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/Design/EmptyCatchBlock.php]: You might add yourself to this 😼

File: src/test/php/PHPMD/Rule/Design/EmptyCatchBlockTest.php [hunk 16]
```diff
@@ -0,0 +1,112 @@
+<?php
+/**
+ * This file is part of PHP Mess Detector.
+ *
+ * Copyright (c) 2008-2017, Manuel Pichler <mapi@phpmd.org>.
```
Comment (ravage84) [inline src/test/php/PHPMD/Rule/Design/EmptyCatchBlockTest.php]: Old file header doc block

File: src/test/php/PHPMD/Rule/Design/EmptyCatchBlockTest.php [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+ * @copyright 2008-2017 Manuel Pichler. All rights reserved.
+ * @license   http://www.opensource.org/licenses/bsd-license.php BSD License
+ */
+
+namespace PHPMD\Rule\Design;
+
+use PHPMD\AbstractTest;
+
+/**
+ * @author    Grégoire Paris <postmaster@greg0ire.fr>
```
Comment (ravage84) [inline src/test/php/PHPMD/Rule/Design/EmptyCatchBlockTest.php]: Add class short description: ``Empty Catch Block Test``.

File: src/test/resources/files/Rule/Design/EmptyCatchBlock/testRuleAppliesToFunctionWithEmptyCatchBlock.php [hunk 18]
```diff
@@ -0,0 +1,17 @@
+<?php
+
```
Comment (ravage84) [inline src/test/resources/files/Rule/Design/EmptyCatchBlock/testRuleAppliesToFunctionWithEmptyCatchBlock.php]: Add new file header doc block.

File: src/test/resources/files/Rule/Design/EmptyCatchBlock/testRuleNotAppliesToCatchBlockWithComments.php [hunk 19]
```diff
@@ -0,0 +1,32 @@
+<?php
+
```
Comment (ravage84) [inline src/test/resources/files/Rule/Design/EmptyCatchBlock/testRuleNotAppliesToCatchBlockWithComments.php]: Add new file header doc block.

File: src/test/resources/files/Rule/Design/EmptyCatchBlock/testRuleNotAppliesToFunctionWithNonEmptyCatchBlock.php [hunk 20]
```diff
@@ -0,0 +1,10 @@
+<?php
+
```
Comment (ravage84) [inline src/test/resources/files/Rule/Design/EmptyCatchBlock/testRuleNotAppliesToFunctionWithNonEmptyCatchBlock.php]: Add new file header doc block.

File: src/test/resources/files/Rule/Design/EmptyCatchBlock/testRuleNotAppliesToMethodWithoutTryCatchBlock.php [hunk 21]
```diff
@@ -0,0 +1,17 @@
+<?php
+
```
Comment (ravage84) [inline src/test/resources/files/Rule/Design/EmptyCatchBlock/testRuleNotAppliesToMethodWithoutTryCatchBlock.php]: Add new file header doc block.

File: src/test/resources/files/Rule/Design/EmptyCatchBlock/testRuleWorksWithNestedTryCatchBlocksAndNonSPLExceptions.php [hunk 22]
```diff
@@ -0,0 +1,17 @@
+<?php
+
```
Comment (ravage84) [inline src/test/resources/files/Rule/Design/EmptyCatchBlock/testRuleWorksWithNestedTryCatchBlocksAndNonSPLExceptions.php]: Add new file header doc block.

File: src/test/php/PHPMD/Rule/Design/EmptyCatchBlockTest.php [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+     */
+    public function testRuleNotAppliesToCatchBlockWithComments()
+    {
+        $rule = new EmptyCatchBlock();
+        $rule->setReport($this->getReportMock(0));
+        $rule->apply($this->getFunction());
+    }
+
+    /**
+     * testRuleNotAppliesToCatchBlockWithComments
```
Comment (ravage84) [inline src/test/php/PHPMD/Rule/Design/EmptyCatchBlockTest.php]: testRuleWorksWithNestedTryCatchBlocksAndNonSPLExceptions

**Comment by greg0ire** [comment]
Thanks for carrying this @eeree !

**Comment by ravage84** [comment]
Lovely! Keep up the good work!

**Comment by eeree** [comment]
> The file header doc blocks aren't the new ones but I'm going to merge and fix it myself.

Thanks a lot!

**Comment by ravage84** [comment]
Fixed in https://github.com/phpmd/phpmd/pull/481/commits/4d750f1c17e6a6ca14042df0d591355efce167a1

**Pull request merged**